### PR TITLE
mds: try to find the ino in the peer in case the inode is exported

### PR DESCRIPTION
**Pull request by lxbsz**
The inode maybe exported to another rank during the requests being retried, we need to retry to fetch or recover it from other peer and then forward it to the auth mds.

Fixes: https://tracker.ceph.com/issues/67358

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by rishabh-d-dave — CHANGES_REQUESTED** [review]

File: src/mds/Server.cc [hunk 1]
```diff
@@ -5696,10 +5696,20 @@ bool Server::xlock_policylock(const MDRequestRef& mdr, CInode *in, bool want_lay
...
     respond_to_request(mdr, -CEPHFS_ESTALE);
     return nullptr;
   }
+  /*
+   * The inode maybe exported to another rank, we need to retry to
+   * find the ino and then forward it to the auth mds.
+   */
+  if (!in) {
+    dout(10) << "fail on cephfs_estale but attempting recovery" << dendl;
+    inodeno_t ino = refpath.get_ino();
```
[the remark this answers is not in the document]
Comment (rishabh-d-dave) [reply src/mds/Server.cc:5709]: `make check` failed for same reason.

[the remark this answers is not in the document]
Comment (lxbsz) [reply src/mds/Server.cc:5709]: Actually we can just remove this line and use the `ino` from the parameter.

**Comment by lxbsz** [comment]
jenkins test api

File: src/mds/Server.cc [hunk 2]
```diff
@@ -5696,10 +5696,19 @@ bool Server::xlock_policylock(const MDRequestRef& mdr, CInode *in, bool want_lay
 CInode* Server::try_get_auth_inode(const MDRequestRef& mdr, inodeno_t ino)
 {
   CInode *in = mdcache->get_inode(ino);
-  if (!in || in->state_test(CInode::STATE_PURGING)) {
+  if (in && in->state_test(CInode::STATE_PURGING)) {
     respond_to_request(mdr, -CEPHFS_ESTALE);
     return nullptr;
   }
+  /*
```
Comment (vshankar) [inline src/mds/Server.cc:5703]: nit: possibly move this code block above the `if (in && .._)`. That would avoid checking `in` again.

Comment (lxbsz) [reply src/mds/Server.cc:5703]: Sure, will fix it.

**Pull request closed without merging**